### PR TITLE
Alexnet Example with Imagenet

### PR DESCRIPTION
**Pull request by Wapaul1**
Actual network works with random input data and labels, needs some work to input ImageNet however.

### CODE REVIEW

File: examples/imagenet_cnn/functions.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
+import tensorflow as tf
+import numpy as np
+import ray
+
+@ray.remote(16 * [np.ndarray], [])
+def updateweights(*weight):
```
Comment (pcmoritz) [inline examples/imagenet_cnn/functions.py]: can you rename to update_weights?

File: examples/imagenet_cnn/functions.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+  feed_dict = dict(zip(placeholders, weight))
+  sess.run(assignment, feed_dict=feed_dict)
+
+@ray.remote([np.ndarray, np.ndarray], 16 * [np.ndarray])
+def computegrad(X, Y):
```
Comment (pcmoritz) [inline examples/imagenet_cnn/functions.py]: can you rename to compute_grad?

File: examples/imagenet_cnn/functions.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  return sess.run([g for (g,v) in compgrads], feed_dict={images:X, y_true:Y, dropout:0.5})
+
+def printaccuracy(X, Y):
```
Comment (pcmoritz) [inline examples/imagenet_cnn/functions.py]: can you rename to print_accuracy?

File: examples/imagenet_cnn/functions.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    print sess.run(accuracy, feed_dict={images:X, y_true:Y, dropout:1.0})
+
+def setupvariables(kernelshape, biasshape):
```
Comment (pcmoritz) [inline examples/imagenet_cnn/functions.py]: can you rename to setup_variables?

File: examples/imagenet_cnn/functions.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+
+images = tf.placeholder(tf.float32, shape=[None, 224, 224, 3])
+y_true = tf.placeholder(tf.float32, shape=[None, 1000])
+parameters = []
+assignment = []
+placeholders = []
+# conv1
+with tf.name_scope('conv1') as scope:
+  newvars = setupvariables([11, 11, 3, 96], [96])
```
Comment (pcmoritz) [inline examples/imagenet_cnn/functions.py]: can you say param, vars, assig = setupvariables(...) and then replace parameters[-x] below by the respective variable? That would be clearer I think.

File: examples/imagenet_cnn/driver.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+s3 = boto3.resource("s3")
+imagenet_bucket = s3.Bucket(args.s3_bucket)
+objects = imagenet_bucket.objects.filter(Prefix=args.key_prefix)
+images = [obj.key for obj in objects.all()]
+
+imagenet = ray.get(imagenet.load_tarfiles_from_s3(args.s3_bucket, map(str, images), [256, 256]))
+print imagenet
+X = map(lambda img: img[0], imagenet)
+
+s5 = boto3.client("s3")
```
Comment (pcmoritz) [inline examples/imagenet_cnn/driver.py]: can reuse the variable from above

File: examples/imagenet_cnn/driver.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+labels = s5.get_object(Bucket=args.s3_bucket, Key=args.label_file)
+lines = labels["Body"].read().split("\n")
+print "lines\n"
+imagepairs = map(lambda line: line.split(" ", 2), lines)
+imagepairs = ray.put(dict(map(lambda tup: (re.sub("(.+)/(.+)", r"\2", tup[0]), tup[-1]), imagepairs)))
```
Comment (pcmoritz) [inline examples/imagenet_cnn/driver.py]: use basename from https://docs.python.org/2/library/os.path.html

File: examples/imagenet_cnn/driver.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+import functions
+
+
+numofshuffles = 5
+num_workers = 3
+batchnum = 0
+weights = []
+
+parser = argparse.ArgumentParser(description="Parse information for data loading.")
+parser.add_argument("--s3-bucket", default="sparknet", type=str, help="Name of the bucket that contains the image data.")
```
Comment (robertnishihara) [inline examples/imagenet_cnn/driver.py]: There should be no default `s3-bucket` (sparknet only works for us). Also, this field should be required.

File: examples/imagenet_cnn/driver.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  print("Weights recieved")
+  weightrefs = map(ray.put, weights)
+  print("Weightrefs created")
+  for i in range(num_workers):
+    functions.update_weights(*weightrefs)
+  if (batchnum % 100 == 0):
+    temp = random.choice(batches)
+    xref = temp[0]
+    yref = temp[1]
+    print ray.get(functions.print_accuracy(xref, yref))
```
Comment (robertnishihara) [inline examples/imagenet_cnn/driver.py]: You can probably make the above a single line
```python
xref, yref = random.choice(batches)
```

File: examples/imagenet_cnn/functions.py [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+  """
+  toshuffle = zip(images1, labels1) + zip(images2, labels2)
+  np.random.shuffle(toshuffle)
+  length = len(images1)
+  firstref = zip(*toshuffle[0:length])
+  shuffledarray = map(np.asarray, firstref)
+  return shuffledarray[0], shuffledarray[1]
+
+@ray.remote([Tuple[Tuple, Tuple]], [list])
+def shufflestuples(zippedimages):
```
Comment (robertnishihara) [inline examples/imagenet_cnn/functions.py]: take two argument instead of 1

File: examples/imagenet_cnn/functions.py [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+    images1 (ndarray): Images of first batch
+    labels1 (ndarray): Labels of first batch
+    images2 (ndarray): Images of second batch
+    labels2 (ndarray): Labels of second batch
+  
+  Returns:
+    ndarray: Shuffled images for first batch
+    ndarray: Shuffled images for first batch
+  """
+  toshuffle = zip(images1, labels1) + zip(images2, labels2)
```
Comment (robertnishihara) [inline examples/imagenet_cnn/functions.py]: use _ to separate words in variable names

File: examples/imagenet_cnn/driver.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+import os
+import tensorflow as tf
+import imagenet
+import argparse
+import boto3
+import random
+
+import functions
+
+
```
Comment (robertnishihara) [inline examples/imagenet_cnn/driver.py]: two newlines

File: examples/imagenet_cnn/driver.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+numofshuffles = 5
```
Comment (robertnishihara) [inline examples/imagenet_cnn/driver.py]: rename num_shuffles

File: examples/imagenet_cnn/driver.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+labels = s5.get_object(Bucket=args.s3_bucket, Key=args.label_file)
+lines = labels["Body"].read().split("\n")
+imagepairs = map(lambda line: line.split(" ", 2), lines)
+imagepairs = ray.put(dict(map(lambda tup: (os.path.basename(tup[0]), tup[-1]), imagepairs)))
+imagenames = map(lambda img: img[1], imagenet)
+Y = map(lambda x: functions.convert(x, imagepairs), imagenames)
+batches = zip(X,Y)
+for i in range(numofshuffles):
+  newshuffle = np.random.permutation(batches)
+  batches = map(lambda tup:ray.get(functions.shufflestuples(tup)), zip(newshuffle,batches))
```
Comment (robertnishihara) [inline examples/imagenet_cnn/driver.py]: newshuffle -> permuted_batches

File: examples/imagenet_cnn/driver.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+  print ray.get(batches[0][1])
+print("Imagenet downloaded")
```
Comment (robertnishihara) [inline examples/imagenet_cnn/driver.py]: please explain what is happening everywhere

File: examples/imagenet_cnn/driver.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+s3 = boto3.resource("s3")
+imagenet_bucket = s3.Bucket(args.s3_bucket)
+objects = imagenet_bucket.objects.filter(Prefix=args.key_prefix)
+images = [obj.key for obj in objects.all()]
+
+imagenet = ray.get(imagenet.load_tarfiles_from_s3(args.s3_bucket, map(str, images), [256, 256]))
+meanref = functions.compute_mean_image(imagenet)
+X = map(lambda img: img[0], imagenet)
+s3 = boto3.client("s3")
+labels = s5.get_object(Bucket=args.s3_bucket, Key=args.label_file)
```
Comment (robertnishihara) [inline examples/imagenet_cnn/driver.py]: s5

**Comment by robertnishihara** [comment]
Replaced by #319.

**Pull request closed without merging**